crates/lib/src/qpu/api.rs

```diff
@@ -356,6 +326,56 @@ impl ExecutionOptions {
     }
 }
 
+#[cached(
+    result = true,
+    time = 3600,
```

Comment:
@kalzoo How often would an accessor change?

I think ideally we'd just keep the accessor cached indefinitely but force a cache update any time we can't connect to a cached accessor. I don't think that can be easily done without losing the benefit of this convenient proc-macro, though.

Reply:
It looks like the proc macro generates a function that can be used to refresh the cache. We might be able to utilize it to recover from a connection error? That might be overly complicated to try and do at this point, but it might be the most robust way of doing it. I can't link to the specific example, but it's [one of the last ones in these docs](https://docs.rs/cached/0.44.0/cached/proc_macro/index.html).

As I write this, I wonder that, even if we could transparently switch accessors in the client, I wonder if it will still be disruptive to the user if they have jobs queued on the old accessor that haven't been processed yet. So maybe we shouldn't even bother trying to switch accessors in a given session?

Reply:
Oh, interesting, it just generates a function ending with `_prime_cache` in the same namespace? _Surely_ that's documented somewhere, not just implicitly in those examples... but yes, that seems like a decent approach in case we end up needing manual control.

I think I need to better understand, conceptually, what an "accessor" is here. Kalan confirmed above that we do indeed expect to have multiple accessors for one QPU ID, but what does that _mean_? Are they all "pointing" to the same ConServ instance, or not?

If not, then you're right that changing accessors would be disruptive; my testing in this branch showed that, without caching, the `accessor` endpoint is hit twice for each run, presumably once to submit the job and once to get the results.

Reply:
(I've checked Confluence a few times to be sure I'm not missing something obvious, and...well, if I'm missing something, it's definitely not _obvious_. And the API documentation itself doesn't really indicate to me what "accessors" are conceptually.)

Reply:
You're right that this is lightly documented, and that stands to be improved in our public docs.

This is about service discovery, and accessors are the advertised means of accessing the particular QPU. They're not pointing to the same service, but to the same QPU (architecture). This is one example of how there may be many:

* one direct access, gRPC API (what we have today)
* an alternative future ™️ API if and when we roll that out
* one or more gateways, where each offers lowest-latency access from different geographical areas (where we can accelerate access via some private backbone as an alternative to public internet - imagine AUS or UK gateways for a US QPU)
* simulated endpoints where that architecture is available but backed by something that isn't a QPU (and of course, marked as such)

Today, accessors are simple: gateway and direct access, that's it (in practice). But the mechanism is meant to be general enough to serve its purpose well into the future, and let us evolve access to our QPUs without breaking changes or new API routes.

Reply:
As far as the TTL length, that's a good point that changing the accessor, even from one valid-and-working accessor to another, may break a client's workflow.

We can't cache indefinitely because there may be long-running processes which rely on this code, and those would never pick up changes without a bounce (not an ideal requirement).

I'd err on the side of shorter right now, imagining the scenario where we publish data that results in a broken client and then need to roll back or fix. How long of downtime would we want to accept in that case? Some clients may fail and be restarted (thus flushing the cache), but any that have retries built in would not.

If we can apply Kyle's suggestion to flush the cache in case of errors, then it becomes easier to accept an hour's wait. However, that does add complexity, either in that we force a request retry (which may not be idempotent) or we require the user to do so.

Reply:
I've gone back to 1 minute. I'll open a separate ticket for ensuring that clients are always able to get results even if accessors change.

Reply:
https://github.com/rigetti/qcs-sdk-rust/issues/341